include/learn/dataset_view.h

```diff
@@ -0,0 +1,230 @@
+/**
+ * @file dataset_view.h
+ * @author Chase Geigle
+ *
+ * All files in META are released under the MIT license. For more details,
+ * consult the file LICENSE in the root of the project.
+ */
+
+#ifndef META_LEARN_DATASET_VIEW_H_
+#define META_LEARN_DATASET_VIEW_H_
+
+#include <iostream>
+#include <iterator>
+#include "learn/dataset.h"
+#include "util/comparable.h"
+#include "util/random.h"
+
+namespace meta
+{
+namespace learn
+{
+
+/**
+ * A non-owning view of a dataset. This is typically what the learning
+ * algorithms will see: their own view of a dataset, which they can
+ * shuffle/permute as they see fit. Derived classes of dataset_view will
+ * typically also define the label value for each element in the dataset
+ * that returns something appropriate for that kind of algorithm: +1 or -1
+ * for binary classification, a class_label for multiclass, a double for
+ * regression, etc.
+ */
+class dataset_view
+{
+  public:
+    using instance_type = dataset::instance_type;
+    using size_type = dataset::size_type;
+
+    class iterator;
+
+    dataset_view(const dataset& dset)
+        : dataset_view{dset, std::mt19937_64{std::random_device{}()}}
+    {
+        // nothing
+    }
+
+    template <class RandomEngine>
+    dataset_view(const dataset& dset, RandomEngine&& rng)
+        : dset_{&dset},
+          indices_(dset.size()),
+          rng_(std::forward<RandomEngine>(rng))
+    {
+        std::iota(indices_.begin(), indices_.end(), 0);
+    }
+
+    // subset constructor
+    dataset_view(const dataset_view& dv, iterator first, iterator last)
+        : dset_{dv.dset_}, rng_{dv.rng_}
+    {
+        indices_.reserve(std::distance(first, last));
+        for (; first != last; ++first)
+            indices_.emplace_back(first.index());
+    }
+
+    void shuffle()
```

Comment:
should the implementation of this function be part of the `random` namespace (e.g. taking two `random_access_iterator`s)? I can see myself wanting to use it to shuffle other things deterministically.